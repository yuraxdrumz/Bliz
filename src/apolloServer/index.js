export {
  runQuery,
  LogFunction,
  LogMessage,
  LogStep,
  LogAction
} from './runQuery'
export { runHttpQuery, HttpQueryError } from './runHttpQuery'
export { resolveGraphqlOptions } from './graphqlOptions'
