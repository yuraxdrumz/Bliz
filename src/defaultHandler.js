let defaultHandler = function(req ,res, err){
  if(err){
    res.statusCode = err.status || 500
    res.json({error:err.toString()})

  }else{
    res.statusCode = 404
    res.json({error:`${req.method.toUpperCase()} - ${req.url} not found...`})
  }
}
export default defaultHandler