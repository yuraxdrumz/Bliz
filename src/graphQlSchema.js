import {
    CreateArray,
    AssignHandler,
    GetObjProps,
    CreateObjectArray
  } from './objectFactories'
  import { checkBaseUtil } from './utils'
  
  const GraphQlSchemaCreator = name => {
    const _graphql = {}
    const pathData = {
        name,
        type: ``,
        resolver: {}
    }
    return Object.assign(
        _graphql,
        AssignHandler('type', pathData, _graphql),
        AssignHandler('resolver', pathData.resolver, _graphql, true),
        GetObjProps(pathData),
    )
  }
  
  export default GraphQlSchemaCreator