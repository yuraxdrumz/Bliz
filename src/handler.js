// main function
function createHandler (request, response, defaultHandler, midHandler, Joi, urlUtil, handleNestedRoutersUtil,populateParamsUtil, populateQueryUtil, populateUrlOptions, middleWares, routes, injected, app, Promise) {
  // receive all middlewares from routers and apps if exist and concatanate them
  middleWares = middleWares.reduce((prev, curr) => prev.concat(curr), [])
  // handler to be passed to http.createServer
  async function handler(req,res){
    // set proto of req and res to point to our req and res
    req.__proto__ = request
    res.__proto__ = response
    if(!req.params) req.params = {}
    if(!req.query) req.query = {}
    // get url parts
    let { method, splitRest } = urlUtil(req.url, req.method)
    // check all url combinations possible
    const urlCombinationOptions = populateUrlOptions(splitRest)
    // get all middlewares collected on routers existing hit on request
    const { baseOfRequest, rest, combinedRoutersMids } = handleNestedRoutersUtil(urlCombinationOptions, routes)
    // populate req.query and return the final url to check on routers object
    const finalRest = populateQueryUtil(req, rest) || rest
    // populate req.params and return bool if need to skip param check or not
    const { param, canSkipBecauseParams } = populateParamsUtil(req, routes, baseOfRequest, method, finalRest)
    // global middleware, if exists work with it, if throws error go to global handler
    // check routers middleware
    try {
      // if global middlewares exists, execute them
      if (middleWares){
        await midHandler(Promise, req, res, middleWares)
      }
      // if routers middlewares exist, execute them
      if (combinedRoutersMids) {
        await midHandler(Promise, req, res, combinedRoutersMids)
      }
    } catch (middleWareError) {
      // in case of error in middlewares call defualt handler
      return defaultHandler(req, res, middleWareError)
    }
    // something is not defined go to default handler
    if(!canSkipBecauseParams){
      if (!routes[baseOfRequest])return defaultHandler(req, res)
      if (!routes[baseOfRequest][method])return defaultHandler(req, res)
      if (!routes[baseOfRequest][method][finalRest]) return defaultHandler(req, res)
    }
    // current route after all checks
    const currentRoute = routes[baseOfRequest][method][!canSkipBecauseParams ? finalRest : param ].getObjProps()
    // try router middleware => route middleware=> route handler=> if err check route err handler=>
    // if err in err handler or err handler not exists => router err handler => if not go to global handler
    try {
      const { middleWareArr, describe, handler } = currentRoute
      // if route middleware, execute
      if (middleWareArr && middleWareArr.length > 0){
        await midHandler(Promise, req, res, middleWareArr)
      }
      // if validation schemes exist, execute them
      if(describe && describe.requests && describe.requests.length > 0){
        for(let i=0; i<describe.requests.length; i++){
          describe.requests[i].schema(req[describe.requests[i].in])
        }
      }
      // call handler with req, res and injected object from app.inject      
      handler(req, res, injected)
    } catch (errorFromHandler) {
      // here, it is the same as with middlewares but backwards, try route err handler, next up try router err handler and finally try global middleware
      try{
        if(currentRoute.errHandler){
          currentRoute.errHandler(req, res, errorFromHandler)
        }else if(routes[baseOfRequest].routerErrorHandler){
          routes[baseOfRequest].routerErrorHandler(req, res, errorFromHandler)
        }else{
          defaultHandler(req, res, errorFromHandler)
        }
      }catch(errorFromErrorHandlers){
        defaultHandler(req, res, errorFromErrorHandlers)
      }

    }
  }
  return { handler }
}

export default createHandler
