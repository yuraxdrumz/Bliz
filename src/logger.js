// throw err and parse it to get file name and line num
let colors        = ['\x1b[31m','\x1b[32m','\x1b[33m','\x1b[35m','\x1b[36m'];
let underline     = '\x1b[4m';
const getStack = () => {
  try {
    throw new Error('throwing this to get info of caller')
  } catch (e) {
    const stack = e.stack.split('\n').slice(1)[2]
    const filename = stack.substr(stack.indexOf('(')).replace('(', '').replace(')', '')
    return filename
  }
}

// main function
const log = (getStack) => (...args) => {
  const random = Math.floor(Math.random() * colors.length)
  const prefix = underline + ' ' + colors[random] + getStack() + '\x1b[0m'
  switch (args.length) {
    case 0:
      console.log(`${prefix} no arguments were passed! \n`)
      break
    default:
      args.forEach(each => {
        each instanceof Error ? console.error(`${prefix} ${each.stack}`) : console.log(`${prefix} ${each}`)
      })
      break
  }
}

export default log(getStack)
