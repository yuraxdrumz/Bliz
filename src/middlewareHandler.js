// handler for express middlewares with next...
async function midHandler(Promise, req, res, arr){
  function next(resolve, reject, ...args){
    if(args.length > 0) return reject(args[0])
    return resolve()
  }
  for(let item of arr){
    await new Promise((resolve, reject)=>{
      item(req,res,next.bind(this,resolve, reject))
    })
  }
}

export default midHandler