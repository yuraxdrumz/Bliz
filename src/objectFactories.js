// receive an http and a handler and return a listen func
const Listen = (name, handlerFactory, http) => ({
  [name]: (...args) => {
    const { handler } = handlerFactory()
    const server = http.createServer(handler)
    return server.listen.apply(server, args)
  }
})


const PrettyPrint = (treeifyDep, entity, chainLink) =>({
  prettyPrint: log =>{
    let logger = console.log
    let shortEntity = {}
    const keysOfEntity = Object.keys(entity)

    if(log && typeof log === 'function') logger = log

    for(let key of keysOfEntity){
      let obj = {}
      let options = ['get','post','put','del']
      for(let option of options){
        let routeValues = Object.keys(entity[key][option])
        if(routeValues.length > 0){
          let routeKey = option.toUpperCase()
          let value = {}
          for(let route of routeValues){
            value[route] = ''
            const assignedOption = {[routeKey]:value}
            Object.assign(obj,assignedOption)
            shortEntity[key] = obj

          }
        }
      }
    }
    logger(treeifyDep.asTree(shortEntity))
    return chainLink
  }
})

// method creator for router
const Method = (name, object,chainLink) => ({
  [name]: data =>{
    object[name][data.getObjProps().path] = data
    return chainLink
  }
})


// assign creator
const AssignHandler = (name, object, chainLink) =>({
  [name]:data =>{
    object[name] = data
    return chainLink
  }
})


// list object properties
const GetObjProps = obj =>({
  getObjProps:()=>(obj)
})

const EventsCreator = eventEmitter =>({
  events:new eventEmitter({wildcard:true})
})

// when called, receives an object
// returns new object
const CreateNewObjOf = (name, obj, ...deps) => ({
  [`create${name}`]: data =>{
    return Object.assign(
      {},
      obj(data, deps)
    )
  }
})
// pushes data to array
const CreateArray = (name, arr, chainLink) => ({
  [name]:data=>{
    arr.push(data)
    return chainLink
  }
})


export {
  AssignHandler,
  CreateArray,
  CreateNewObjOf,
  Method,
  GetObjProps,
  Listen,
  PrettyPrint,
  EventsCreator
}