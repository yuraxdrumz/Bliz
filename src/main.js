import RouterCreator from './router'
import PathCreator from './path'
import { Listen, CreateArray, CreateNewObjOf, GetObjProps, PrettyPrint, EventsCreator } from './objectFactories'
import { urlUtil, populateQueryUtil,populateRoutersUtil,populateParamsUtil, handleNestedRoutersUtil, populateUrlOptions, populateSubAppsUtil } from './utils'
import defaultHandler from './defaultHandler'
import midHandler from './middlewareHandler'
import http from 'http'
import RegisterRouters from './registerRouters'
import CreateHandler from './handler'
import request from './request'
import response from './response'
import treeify from 'treeify'
import Joi from 'joi'
import EventEmitter from 'eventemitter2'
import Promise from 'bluebird'

// main instance creator, returns an instance of bliz app
const BlizApp = (request, response, Joi, RouterCreator, Listen, defaultHandler, midHandler, PathCreator, http, urlUtil, populateRoutersUtil, handleNestedRoutersUtil,populateParamsUtil, populateQueryUtil, populateUrlOptions, createHandler, GetObjProps, populateSubAppsUtil, treeify, EventsCreator, EventEmitter, Promise) => {
  const _Instance = {}
  const _middleWares = []
  const _routersObject = {}
  const _createHandler = CreateHandler.bind(this,request, response ,defaultHandler, midHandler, Joi, urlUtil, handleNestedRoutersUtil,populateParamsUtil, populateQueryUtil, populateUrlOptions, _middleWares, _routersObject, _Instance, Promise)
  const _subApps = []
  return Object.assign(
    _Instance,
    PrettyPrint(treeify, _routersObject, _Instance),
    CreateNewObjOf('Router', RouterCreator, treeify),
    RegisterRouters(populateRoutersUtil, populateSubAppsUtil, _middleWares, _routersObject, _subApps, _Instance),
    CreateArray('middleware',_middleWares, _Instance),
    CreateArray('subApp', _subApps, _Instance),
    CreateNewObjOf('Path', PathCreator, treeify),
    EventsCreator(EventEmitter),
    GetObjProps({_middleWares, _routersObject, _subApps}),
    Listen('listen', _createHandler, http)
  )
}

const BlizCreator = () => {
  return Object.assign(
    {},
    BlizApp(
      request,
      response,
      Joi,
      RouterCreator,
      Listen,
      defaultHandler,
      midHandler,
      PathCreator,
      http,
      urlUtil,
      populateRoutersUtil,
      handleNestedRoutersUtil,
      populateParamsUtil,
      populateQueryUtil,
      populateUrlOptions,
      CreateHandler,
      GetObjProps,
      populateSubAppsUtil,
      treeify,
      EventsCreator,
      EventEmitter,
      Promise
    )
  )
}

export default BlizCreator
export { request, response, Joi }