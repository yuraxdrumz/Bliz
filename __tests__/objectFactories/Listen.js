import { Listen } from '../../src/objectFactories'

describe('Listen suite', () => {
  test('Should return object with method on it, method name should be as 1st param passed', ()=>{
    const name = 'random name'
    expect(Listen(name)).toMatchObject({[name]:expect.any(Function)})
  })
  test('Should extract handler from handlerFactory, pass it to http.createServer and return server.listen', ()=>{
    const name = 'listenName'
    const handlerFactory = jest.fn(()=>({handler:jest.fn()}))
    const listen = jest.fn((...args)=>args)
    const createServer = jest.fn((handler)=>({listen}))
    const http = {
      createServer
    }
    expect(Listen(name, handlerFactory, http)[name](3000, ()=>console.log('running on port 3000'))).toMatchSnapshot()
    expect(handlerFactory).toHaveBeenCalled()
    expect(createServer).toHaveBeenCalled()
    expect(createServer.mock.calls[0][0]).toEqual(expect.any(Function))
    expect(listen).toHaveBeenCalled()
    expect(listen.mock.calls[0]).toMatchSnapshot()

  })
})