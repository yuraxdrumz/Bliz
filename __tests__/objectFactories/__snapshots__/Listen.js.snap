// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listen suite Should extract handler from handlerFactory, pass it to http.createServer and return server.listen 1`] = `
Array [
  3000,
  [Function],
]
`;

exports[`Listen suite Should extract handler from handlerFactory, pass it to http.createServer and return server.listen 2`] = `
Array [
  3000,
  [Function],
]
`;
